### PR TITLE
Maint/minor improvements

### PR DESCRIPTION
**Pull request by mfeurer**

### CODE REVIEW

File: openml/flows/flow.py [hunk 1]
```diff
@@ -51,10 +52,16 @@ def _generate_flow_xml(self):
             Flow represented as XML string.
         """
         model = self.model
+
+        if config._testmode:
+            flow_name = '%s%s' % (config.testsentinel, self.name)
```
Comment (amueller) [inline openml/flows/flow.py]: I don't like this. There should be no testing code in the main code. Why do you want this?

File: openml/flows/flow.py [hunk 2]
```diff
@@ -137,6 +144,12 @@ def _check_flow_exists(name, version):
     if not (type(version) is str and len(version) > 0):
         raise ValueError('Argument \'version\' should be a non-empty string')
 
+    if config._testmode:
```
Comment (amueller) [inline openml/flows/flow.py]: same thing. This should be handled in testing.

File: openml/datasets/dataset.py [hunk 3]
```diff
@@ -75,7 +76,7 @@ def __init__(self, id=None, name=None, version=None, description=None,
                 logger.debug("Data pickle file already exists.")
             else:
                 try:
-                    data = self._get_arff()
```
Comment (amueller) [inline openml/datasets/dataset.py]: why is ``format`` an attribute? Can we find that out from the file? Shouldn't the file reader handle this?

File: openml/tasks/functions.py [hunk 4]
```diff
@@ -235,7 +236,7 @@ def get_task(task_id):
 
     # TODO look into either adding the class labels to task xml, or other
     # way of reading it.
-    class_labels = dataset._retrieve_class_labels()
+    class_labels = dataset._retrieve_class_labels(task.target_feature)
```
Comment (amueller) [inline openml/tasks/functions.py]: why is this called ``target_feature``? That's not a feature, is it? ``target_value``? ``target_name``? ``target_column``?

**Comment by amueller** [comment]
The IO stuff is for python2.x, right? Can you put this in it's own PR and if it passes tests we merge?

**Comment by amueller** [comment]
It seems a bit like each of the commits should be it's own PR ;)

**Comment by amueller** [comment]
can you explain the sentinel business?

File: tests/datasets/test_datasets.py [hunk 5]
```diff
@@ -206,6 +206,15 @@ def test_publish_dataset(self):
...
         self.assertEqual(return_code, 200)
 
+    def test__retrieve_class_labels(self):
+        openml.config.set_cache_directory(self.static_cache_dir)
+        labels = openml.datasets.get_dataset(2)._retrieve_class_labels()
+        self.assertEqual(labels, ['1', '2', '3', '4', '5', 'U'])
+        labels = openml.datasets.get_dataset(2)._retrieve_class_labels(
+            target_attribute='product-type')
+        self.assertEqual(labels, ['C', 'H', 'G'])
+        print(labels)
```
Comment (amueller) [inline tests/datasets/test_datasets.py]: no print please.

**Comment by amueller** [comment]
Do you want to split up the PR? Alternatively, I can cherry pick 8314fac8b478bde672aac3633e42dee9c705133b 7a32fd4c01409992b71405ffad0b7dffdf449889 be03ca604851b8e1383f60fda59b337a871cbe15 9449250fecc16e12ec78d415cab53177e5b65419 into develop, and you can try to reorganize the rest? There is some test coverage, some sparse arff stuff, some sentinel stuff and some target value stuff that I think should be separate PRS.

**Comment by mfeurer** [comment]
Okay, let's do it like that. I assumed the changes are small enough and only fixes, so they could be put together.

**Comment by amueller** [comment]
and why is this not a PR onto master?

**Comment by mfeurer** [comment]
master = current 'stable'
develop = the development branch

according to the [gitflow model](http://nvie.com/posts/a-successful-git-branching-model/)

**Comment by amueller** [comment]
sorry I meant "why is this not a PR onto develop"

**Comment by mfeurer** [comment]
This is now wrong; it wasn't because it actually built upon the already merged changes.

**Comment by amueller** [comment]
I cherry picked the simple changes into develop.

File: openml/flows/flow.py (hunk 1 above)
Comment (mfeurer) [inline openml/flows/flow.py]: To change the name to be unique for each test run. I found this to be more reliable than deleting runs on the test server. Would it help to add a method `get_name` which I mock during test time?

File: openml/flows/flow.py (hunk 2 above)
Comment (mfeurer) [inline openml/flows/flow.py]: Would be handled by the suggestion above.

File: openml/datasets/dataset.py (hunk 3 above)
Comment (mfeurer) [inline openml/datasets/dataset.py]: Format is returned by the dataset xml. In the long run there will be more formats than arff and sparse_arff, so I would prefer to not try to infer this. Plus, the current arff reader can't do this.

File: openml/tasks/functions.py (hunk 4 above)
Comment (mfeurer) [inline openml/tasks/functions.py]: Will fix that.

**Comment by mfeurer** [comment]
Is [this comment](https://github.com/openml/openml-python/pull/150#issuecomment-243802154) still up to date, given your cherry-pick to develop?

File: openml/tasks/functions.py (hunk 4 above)
Comment (mfeurer) [inline openml/tasks/functions.py]: This is actually an upstream issue: http://www.openml.org/api_docs/#!/task/get_task_id

Not sure now if I should fix this.

**Comment by mfeurer** [comment]
This PR is now split into #155 #156 #157 #158

**Pull request closed without merging**